{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2023-10-31 16:38:54,149:INFO - HTTP Request: POST https://lhtazigkmlwzrsknbidn.supabase.co/rest/v1/application \"HTTP/1.1 409 Conflict\"\n"
     ]
    },
    {
     "ename": "APIError",
     "evalue": "{'code': '23505', 'details': 'Key (application_id)=(69) already exists.', 'hint': None, 'message': 'duplicate key value violates unique constraint \"application_pkey\"'}",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mAPIError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[1;32mc:\\SMU\\Y3T1\\IS212 Software Project Management\\spm-g5t3\\pytest\\test.ipynb Cell 1\u001b[0m line \u001b[0;36m1\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/SMU/Y3T1/IS212%20Software%20Project%20Management/spm-g5t3/pytest/test.ipynb#W0sZmlsZQ%3D%3D?line=10'>11</a>\u001b[0m client \u001b[39m=\u001b[39m TestClient(app)\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/SMU/Y3T1/IS212%20Software%20Project%20Management/spm-g5t3/pytest/test.ipynb#W0sZmlsZQ%3D%3D?line=12'>13</a>\u001b[0m \u001b[39m# response = client.get(\"/api/application?application_id=1\")\u001b[39;00m\n\u001b[1;32m---> <a href='vscode-notebook-cell:/c%3A/SMU/Y3T1/IS212%20Software%20Project%20Management/spm-g5t3/pytest/test.ipynb#W0sZmlsZQ%3D%3D?line=13'>14</a>\u001b[0m response \u001b[39m=\u001b[39m client\u001b[39m.\u001b[39;49mpost(\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/SMU/Y3T1/IS212%20Software%20Project%20Management/spm-g5t3/pytest/test.ipynb#W0sZmlsZQ%3D%3D?line=14'>15</a>\u001b[0m         \u001b[39m\"\u001b[39;49m\u001b[39m/api/application\u001b[39;49m\u001b[39m\"\u001b[39;49m,\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/SMU/Y3T1/IS212%20Software%20Project%20Management/spm-g5t3/pytest/test.ipynb#W0sZmlsZQ%3D%3D?line=15'>16</a>\u001b[0m         json\u001b[39m=\u001b[39;49m{\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/SMU/Y3T1/IS212%20Software%20Project%20Management/spm-g5t3/pytest/test.ipynb#W0sZmlsZQ%3D%3D?line=16'>17</a>\u001b[0m             \u001b[39m\"\u001b[39;49m\u001b[39mstaff_id\u001b[39;49m\u001b[39m\"\u001b[39;49m: \u001b[39m140944\u001b[39;49m,\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/SMU/Y3T1/IS212%20Software%20Project%20Management/spm-g5t3/pytest/test.ipynb#W0sZmlsZQ%3D%3D?line=17'>18</a>\u001b[0m             \u001b[39m\"\u001b[39;49m\u001b[39mlisting_id\u001b[39;49m\u001b[39m\"\u001b[39;49m: \u001b[39m1\u001b[39;49m,\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/SMU/Y3T1/IS212%20Software%20Project%20Management/spm-g5t3/pytest/test.ipynb#W0sZmlsZQ%3D%3D?line=18'>19</a>\u001b[0m             \u001b[39m\"\u001b[39;49m\u001b[39mapplication_reason\u001b[39;49m\u001b[39m\"\u001b[39;49m: \u001b[39m\"\u001b[39;49m\u001b[39mtest_application.py\u001b[39;49m\u001b[39m\"\u001b[39;49m\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/SMU/Y3T1/IS212%20Software%20Project%20Management/spm-g5t3/pytest/test.ipynb#W0sZmlsZQ%3D%3D?line=19'>20</a>\u001b[0m         }\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/SMU/Y3T1/IS212%20Software%20Project%20Management/spm-g5t3/pytest/test.ipynb#W0sZmlsZQ%3D%3D?line=20'>21</a>\u001b[0m     )\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/SMU/Y3T1/IS212%20Software%20Project%20Management/spm-g5t3/pytest/test.ipynb#W0sZmlsZQ%3D%3D?line=21'>22</a>\u001b[0m \u001b[39mprint\u001b[39m(response)\n",
      "File \u001b[1;32mc:\\Users\\vince\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\starlette\\testclient.py:590\u001b[0m, in \u001b[0;36mTestClient.post\u001b[1;34m(self, url, content, data, files, json, params, headers, cookies, auth, follow_redirects, allow_redirects, timeout, extensions)\u001b[0m\n\u001b[0;32m    568\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mpost\u001b[39m(  \u001b[39m# type: ignore[override]\u001b[39;00m\n\u001b[0;32m    569\u001b[0m     \u001b[39mself\u001b[39m,\n\u001b[0;32m    570\u001b[0m     url: httpx\u001b[39m.\u001b[39m_types\u001b[39m.\u001b[39mURLTypes,\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    587\u001b[0m     extensions: typing\u001b[39m.\u001b[39mOptional[typing\u001b[39m.\u001b[39mDict[\u001b[39mstr\u001b[39m, typing\u001b[39m.\u001b[39mAny]] \u001b[39m=\u001b[39m \u001b[39mNone\u001b[39;00m,\n\u001b[0;32m    588\u001b[0m ) \u001b[39m-\u001b[39m\u001b[39m>\u001b[39m httpx\u001b[39m.\u001b[39mResponse:\n\u001b[0;32m    589\u001b[0m     redirect \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_choose_redirect_arg(follow_redirects, allow_redirects)\n\u001b[1;32m--> 590\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39msuper\u001b[39;49m()\u001b[39m.\u001b[39;49mpost(\n\u001b[0;32m    591\u001b[0m         url,\n\u001b[0;32m    592\u001b[0m         content\u001b[39m=\u001b[39;49mcontent,\n\u001b[0;32m    593\u001b[0m         data\u001b[39m=\u001b[39;49mdata,  \u001b[39m# type: ignore[arg-type]\u001b[39;49;00m\n\u001b[0;32m    594\u001b[0m         files\u001b[39m=\u001b[39;49mfiles,\n\u001b[0;32m    595\u001b[0m         json\u001b[39m=\u001b[39;49mjson,\n\u001b[0;32m    596\u001b[0m         params\u001b[39m=\u001b[39;49mparams,\n\u001b[0;32m    597\u001b[0m         headers\u001b[39m=\u001b[39;49mheaders,\n\u001b[0;32m    598\u001b[0m         cookies\u001b[39m=\u001b[39;49mcookies,\n\u001b[0;32m    599\u001b[0m         auth\u001b[39m=\u001b[39;49mauth,\n\u001b[0;32m    600\u001b[0m         follow_redirects\u001b[39m=\u001b[39;49mredirect,\n\u001b[0;32m    601\u001b[0m         timeout\u001b[39m=\u001b[39;49mtimeout,\n\u001b[0;32m    602\u001b[0m         extensions\u001b[39m=\u001b[39;49mextensions,\n\u001b[0;32m    603\u001b[0m     )\n",
      "File \u001b[1;32mc:\\Users\\vince\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\httpx\\_client.py:1132\u001b[0m, in \u001b[0;36mClient.post\u001b[1;34m(self, url, content, data, files, json, params, headers, cookies, auth, follow_redirects, timeout, extensions)\u001b[0m\n\u001b[0;32m   1111\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mpost\u001b[39m(\n\u001b[0;32m   1112\u001b[0m     \u001b[39mself\u001b[39m,\n\u001b[0;32m   1113\u001b[0m     url: URLTypes,\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m   1125\u001b[0m     extensions: typing\u001b[39m.\u001b[39mOptional[RequestExtensions] \u001b[39m=\u001b[39m \u001b[39mNone\u001b[39;00m,\n\u001b[0;32m   1126\u001b[0m ) \u001b[39m-\u001b[39m\u001b[39m>\u001b[39m Response:\n\u001b[0;32m   1127\u001b[0m \u001b[39m    \u001b[39m\u001b[39m\"\"\"\u001b[39;00m\n\u001b[0;32m   1128\u001b[0m \u001b[39m    Send a `POST` request.\u001b[39;00m\n\u001b[0;32m   1129\u001b[0m \n\u001b[0;32m   1130\u001b[0m \u001b[39m    **Parameters**: See `httpx.request`.\u001b[39;00m\n\u001b[0;32m   1131\u001b[0m \u001b[39m    \"\"\"\u001b[39;00m\n\u001b[1;32m-> 1132\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mrequest(\n\u001b[0;32m   1133\u001b[0m         \u001b[39m\"\u001b[39;49m\u001b[39mPOST\u001b[39;49m\u001b[39m\"\u001b[39;49m,\n\u001b[0;32m   1134\u001b[0m         url,\n\u001b[0;32m   1135\u001b[0m         content\u001b[39m=\u001b[39;49mcontent,\n\u001b[0;32m   1136\u001b[0m         data\u001b[39m=\u001b[39;49mdata,\n\u001b[0;32m   1137\u001b[0m         files\u001b[39m=\u001b[39;49mfiles,\n\u001b[0;32m   1138\u001b[0m         json\u001b[39m=\u001b[39;49mjson,\n\u001b[0;32m   1139\u001b[0m         params\u001b[39m=\u001b[39;49mparams,\n\u001b[0;32m   1140\u001b[0m         headers\u001b[39m=\u001b[39;49mheaders,\n\u001b[0;32m   1141\u001b[0m         cookies\u001b[39m=\u001b[39;49mcookies,\n\u001b[0;32m   1142\u001b[0m         auth\u001b[39m=\u001b[39;49mauth,\n\u001b[0;32m   1143\u001b[0m         follow_redirects\u001b[39m=\u001b[39;49mfollow_redirects,\n\u001b[0;32m   1144\u001b[0m         timeout\u001b[39m=\u001b[39;49mtimeout,\n\u001b[0;32m   1145\u001b[0m         extensions\u001b[39m=\u001b[39;49mextensions,\n\u001b[0;32m   1146\u001b[0m     )\n",
      "File \u001b[1;32mc:\\Users\\vince\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\starlette\\testclient.py:465\u001b[0m, in \u001b[0;36mTestClient.request\u001b[1;34m(self, method, url, content, data, files, json, params, headers, cookies, auth, follow_redirects, allow_redirects, timeout, extensions)\u001b[0m\n\u001b[0;32m    463\u001b[0m url \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mbase_url\u001b[39m.\u001b[39mjoin(url)\n\u001b[0;32m    464\u001b[0m redirect \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_choose_redirect_arg(follow_redirects, allow_redirects)\n\u001b[1;32m--> 465\u001b[0m \u001b[39mreturn\u001b[39;00m \u001b[39msuper\u001b[39;49m()\u001b[39m.\u001b[39;49mrequest(\n\u001b[0;32m    466\u001b[0m     method,\n\u001b[0;32m    467\u001b[0m     url,\n\u001b[0;32m    468\u001b[0m     content\u001b[39m=\u001b[39;49mcontent,\n\u001b[0;32m    469\u001b[0m     data\u001b[39m=\u001b[39;49mdata,  \u001b[39m# type: ignore[arg-type]\u001b[39;49;00m\n\u001b[0;32m    470\u001b[0m     files\u001b[39m=\u001b[39;49mfiles,\n\u001b[0;32m    471\u001b[0m     json\u001b[39m=\u001b[39;49mjson,\n\u001b[0;32m    472\u001b[0m     params\u001b[39m=\u001b[39;49mparams,\n\u001b[0;32m    473\u001b[0m     headers\u001b[39m=\u001b[39;49mheaders,\n\u001b[0;32m    474\u001b[0m     cookies\u001b[39m=\u001b[39;49mcookies,\n\u001b[0;32m    475\u001b[0m     auth\u001b[39m=\u001b[39;49mauth,\n\u001b[0;32m    476\u001b[0m     follow_redirects\u001b[39m=\u001b[39;49mredirect,\n\u001b[0;32m    477\u001b[0m     timeout\u001b[39m=\u001b[39;49mtimeout,\n\u001b[0;32m    478\u001b[0m     extensions\u001b[39m=\u001b[39;49mextensions,\n\u001b[0;32m    479\u001b[0m )\n",
      "File \u001b[1;32mc:\\Users\\vince\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\httpx\\_client.py:814\u001b[0m, in \u001b[0;36mClient.request\u001b[1;34m(self, method, url, content, data, files, json, params, headers, cookies, auth, follow_redirects, timeout, extensions)\u001b[0m\n\u001b[0;32m    799\u001b[0m     warnings\u001b[39m.\u001b[39mwarn(message, \u001b[39mDeprecationWarning\u001b[39;00m)\n\u001b[0;32m    801\u001b[0m request \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mbuild_request(\n\u001b[0;32m    802\u001b[0m     method\u001b[39m=\u001b[39mmethod,\n\u001b[0;32m    803\u001b[0m     url\u001b[39m=\u001b[39murl,\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    812\u001b[0m     extensions\u001b[39m=\u001b[39mextensions,\n\u001b[0;32m    813\u001b[0m )\n\u001b[1;32m--> 814\u001b[0m \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49msend(request, auth\u001b[39m=\u001b[39;49mauth, follow_redirects\u001b[39m=\u001b[39;49mfollow_redirects)\n",
      "File \u001b[1;32mc:\\Users\\vince\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\httpx\\_client.py:901\u001b[0m, in \u001b[0;36mClient.send\u001b[1;34m(self, request, stream, auth, follow_redirects)\u001b[0m\n\u001b[0;32m    893\u001b[0m follow_redirects \u001b[39m=\u001b[39m (\n\u001b[0;32m    894\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mfollow_redirects\n\u001b[0;32m    895\u001b[0m     \u001b[39mif\u001b[39;00m \u001b[39misinstance\u001b[39m(follow_redirects, UseClientDefault)\n\u001b[0;32m    896\u001b[0m     \u001b[39melse\u001b[39;00m follow_redirects\n\u001b[0;32m    897\u001b[0m )\n\u001b[0;32m    899\u001b[0m auth \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_build_request_auth(request, auth)\n\u001b[1;32m--> 901\u001b[0m response \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_send_handling_auth(\n\u001b[0;32m    902\u001b[0m     request,\n\u001b[0;32m    903\u001b[0m     auth\u001b[39m=\u001b[39;49mauth,\n\u001b[0;32m    904\u001b[0m     follow_redirects\u001b[39m=\u001b[39;49mfollow_redirects,\n\u001b[0;32m    905\u001b[0m     history\u001b[39m=\u001b[39;49m[],\n\u001b[0;32m    906\u001b[0m )\n\u001b[0;32m    907\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m    908\u001b[0m     \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m stream:\n",
      "File \u001b[1;32mc:\\Users\\vince\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\httpx\\_client.py:929\u001b[0m, in \u001b[0;36mClient._send_handling_auth\u001b[1;34m(self, request, auth, follow_redirects, history)\u001b[0m\n\u001b[0;32m    926\u001b[0m request \u001b[39m=\u001b[39m \u001b[39mnext\u001b[39m(auth_flow)\n\u001b[0;32m    928\u001b[0m \u001b[39mwhile\u001b[39;00m \u001b[39mTrue\u001b[39;00m:\n\u001b[1;32m--> 929\u001b[0m     response \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_send_handling_redirects(\n\u001b[0;32m    930\u001b[0m         request,\n\u001b[0;32m    931\u001b[0m         follow_redirects\u001b[39m=\u001b[39;49mfollow_redirects,\n\u001b[0;32m    932\u001b[0m         history\u001b[39m=\u001b[39;49mhistory,\n\u001b[0;32m    933\u001b[0m     )\n\u001b[0;32m    934\u001b[0m     \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m    935\u001b[0m         \u001b[39mtry\u001b[39;00m:\n",
      "File \u001b[1;32mc:\\Users\\vince\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\httpx\\_client.py:966\u001b[0m, in \u001b[0;36mClient._send_handling_redirects\u001b[1;34m(self, request, follow_redirects, history)\u001b[0m\n\u001b[0;32m    963\u001b[0m \u001b[39mfor\u001b[39;00m hook \u001b[39min\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_event_hooks[\u001b[39m\"\u001b[39m\u001b[39mrequest\u001b[39m\u001b[39m\"\u001b[39m]:\n\u001b[0;32m    964\u001b[0m     hook(request)\n\u001b[1;32m--> 966\u001b[0m response \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_send_single_request(request)\n\u001b[0;32m    967\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m    968\u001b[0m     \u001b[39mfor\u001b[39;00m hook \u001b[39min\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_event_hooks[\u001b[39m\"\u001b[39m\u001b[39mresponse\u001b[39m\u001b[39m\"\u001b[39m]:\n",
      "File \u001b[1;32mc:\\Users\\vince\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\httpx\\_client.py:1002\u001b[0m, in \u001b[0;36mClient._send_single_request\u001b[1;34m(self, request)\u001b[0m\n\u001b[0;32m    997\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mRuntimeError\u001b[39;00m(\n\u001b[0;32m    998\u001b[0m         \u001b[39m\"\u001b[39m\u001b[39mAttempted to send an async request with a sync Client instance.\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[0;32m    999\u001b[0m     )\n\u001b[0;32m   1001\u001b[0m \u001b[39mwith\u001b[39;00m request_context(request\u001b[39m=\u001b[39mrequest):\n\u001b[1;32m-> 1002\u001b[0m     response \u001b[39m=\u001b[39m transport\u001b[39m.\u001b[39;49mhandle_request(request)\n\u001b[0;32m   1004\u001b[0m \u001b[39massert\u001b[39;00m \u001b[39misinstance\u001b[39m(response\u001b[39m.\u001b[39mstream, SyncByteStream)\n\u001b[0;32m   1006\u001b[0m response\u001b[39m.\u001b[39mrequest \u001b[39m=\u001b[39m request\n",
      "File \u001b[1;32mc:\\Users\\vince\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\starlette\\testclient.py:342\u001b[0m, in \u001b[0;36m_TestClientTransport.handle_request\u001b[1;34m(self, request)\u001b[0m\n\u001b[0;32m    340\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mBaseException\u001b[39;00m \u001b[39mas\u001b[39;00m exc:\n\u001b[0;32m    341\u001b[0m     \u001b[39mif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mraise_server_exceptions:\n\u001b[1;32m--> 342\u001b[0m         \u001b[39mraise\u001b[39;00m exc\n\u001b[0;32m    344\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mraise_server_exceptions:\n\u001b[0;32m    345\u001b[0m     \u001b[39massert\u001b[39;00m response_started, \u001b[39m\"\u001b[39m\u001b[39mTestClient did not receive any response.\u001b[39m\u001b[39m\"\u001b[39m\n",
      "File \u001b[1;32mc:\\Users\\vince\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\starlette\\testclient.py:339\u001b[0m, in \u001b[0;36m_TestClientTransport.handle_request\u001b[1;34m(self, request)\u001b[0m\n\u001b[0;32m    337\u001b[0m     \u001b[39mwith\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mportal_factory() \u001b[39mas\u001b[39;00m portal:\n\u001b[0;32m    338\u001b[0m         response_complete \u001b[39m=\u001b[39m portal\u001b[39m.\u001b[39mcall(anyio\u001b[39m.\u001b[39mEvent)\n\u001b[1;32m--> 339\u001b[0m         portal\u001b[39m.\u001b[39;49mcall(\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mapp, scope, receive, send)\n\u001b[0;32m    340\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mBaseException\u001b[39;00m \u001b[39mas\u001b[39;00m exc:\n\u001b[0;32m    341\u001b[0m     \u001b[39mif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mraise_server_exceptions:\n",
      "File \u001b[1;32mc:\\Users\\vince\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\anyio\\from_thread.py:277\u001b[0m, in \u001b[0;36mBlockingPortal.call\u001b[1;34m(self, func, *args)\u001b[0m\n\u001b[0;32m    264\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mcall\u001b[39m(\n\u001b[0;32m    265\u001b[0m     \u001b[39mself\u001b[39m, func: Callable[\u001b[39m.\u001b[39m\u001b[39m.\u001b[39m\u001b[39m.\u001b[39m, Awaitable[T_Retval] \u001b[39m|\u001b[39m T_Retval], \u001b[39m*\u001b[39margs: \u001b[39mobject\u001b[39m\n\u001b[0;32m    266\u001b[0m ) \u001b[39m-\u001b[39m\u001b[39m>\u001b[39m T_Retval:\n\u001b[0;32m    267\u001b[0m \u001b[39m    \u001b[39m\u001b[39m\"\"\"\u001b[39;00m\n\u001b[0;32m    268\u001b[0m \u001b[39m    Call the given function in the event loop thread.\u001b[39;00m\n\u001b[0;32m    269\u001b[0m \n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    275\u001b[0m \n\u001b[0;32m    276\u001b[0m \u001b[39m    \"\"\"\u001b[39;00m\n\u001b[1;32m--> 277\u001b[0m     \u001b[39mreturn\u001b[39;00m cast(T_Retval, \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mstart_task_soon(func, \u001b[39m*\u001b[39;49margs)\u001b[39m.\u001b[39;49mresult())\n",
      "File \u001b[1;32mc:\\Users\\vince\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\concurrent\\futures\\_base.py:456\u001b[0m, in \u001b[0;36mFuture.result\u001b[1;34m(self, timeout)\u001b[0m\n\u001b[0;32m    454\u001b[0m     \u001b[39mraise\u001b[39;00m CancelledError()\n\u001b[0;32m    455\u001b[0m \u001b[39melif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_state \u001b[39m==\u001b[39m FINISHED:\n\u001b[1;32m--> 456\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m__get_result()\n\u001b[0;32m    457\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[0;32m    458\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mTimeoutError\u001b[39;00m()\n",
      "File \u001b[1;32mc:\\Users\\vince\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\concurrent\\futures\\_base.py:401\u001b[0m, in \u001b[0;36mFuture.__get_result\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    399\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_exception:\n\u001b[0;32m    400\u001b[0m     \u001b[39mtry\u001b[39;00m:\n\u001b[1;32m--> 401\u001b[0m         \u001b[39mraise\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_exception\n\u001b[0;32m    402\u001b[0m     \u001b[39mfinally\u001b[39;00m:\n\u001b[0;32m    403\u001b[0m         \u001b[39m# Break a reference cycle with the exception in self._exception\u001b[39;00m\n\u001b[0;32m    404\u001b[0m         \u001b[39mself\u001b[39m \u001b[39m=\u001b[39m \u001b[39mNone\u001b[39;00m\n",
      "File \u001b[1;32mc:\\Users\\vince\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\anyio\\from_thread.py:217\u001b[0m, in \u001b[0;36mBlockingPortal._call_func\u001b[1;34m(self, func, args, kwargs, future)\u001b[0m\n\u001b[0;32m    214\u001b[0m             \u001b[39melse\u001b[39;00m:\n\u001b[0;32m    215\u001b[0m                 future\u001b[39m.\u001b[39madd_done_callback(callback)\n\u001b[1;32m--> 217\u001b[0m             retval \u001b[39m=\u001b[39m \u001b[39mawait\u001b[39;00m retval\n\u001b[0;32m    218\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_cancelled_exc_class:\n\u001b[0;32m    219\u001b[0m     future\u001b[39m.\u001b[39mcancel()\n",
      "File \u001b[1;32mc:\\Users\\vince\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\fastapi\\applications.py:1115\u001b[0m, in \u001b[0;36mFastAPI.__call__\u001b[1;34m(self, scope, receive, send)\u001b[0m\n\u001b[0;32m   1113\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mroot_path:\n\u001b[0;32m   1114\u001b[0m     scope[\u001b[39m\"\u001b[39m\u001b[39mroot_path\u001b[39m\u001b[39m\"\u001b[39m] \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mroot_path\n\u001b[1;32m-> 1115\u001b[0m \u001b[39mawait\u001b[39;00m \u001b[39msuper\u001b[39m()\u001b[39m.\u001b[39m\u001b[39m__call__\u001b[39m(scope, receive, send)\n",
      "File \u001b[1;32mc:\\Users\\vince\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\starlette\\applications.py:122\u001b[0m, in \u001b[0;36mStarlette.__call__\u001b[1;34m(self, scope, receive, send)\u001b[0m\n\u001b[0;32m    120\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mmiddleware_stack \u001b[39mis\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n\u001b[0;32m    121\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mmiddleware_stack \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mbuild_middleware_stack()\n\u001b[1;32m--> 122\u001b[0m \u001b[39mawait\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mmiddleware_stack(scope, receive, send)\n",
      "File \u001b[1;32mc:\\Users\\vince\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\starlette\\middleware\\errors.py:184\u001b[0m, in \u001b[0;36mServerErrorMiddleware.__call__\u001b[1;34m(self, scope, receive, send)\u001b[0m\n\u001b[0;32m    179\u001b[0m     \u001b[39mawait\u001b[39;00m response(scope, receive, send)\n\u001b[0;32m    181\u001b[0m \u001b[39m# We always continue to raise the exception.\u001b[39;00m\n\u001b[0;32m    182\u001b[0m \u001b[39m# This allows servers to log the error, or allows test clients\u001b[39;00m\n\u001b[0;32m    183\u001b[0m \u001b[39m# to optionally raise the error within the test case.\u001b[39;00m\n\u001b[1;32m--> 184\u001b[0m \u001b[39mraise\u001b[39;00m exc\n",
      "File \u001b[1;32mc:\\Users\\vince\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\starlette\\middleware\\errors.py:162\u001b[0m, in \u001b[0;36mServerErrorMiddleware.__call__\u001b[1;34m(self, scope, receive, send)\u001b[0m\n\u001b[0;32m    159\u001b[0m     \u001b[39mawait\u001b[39;00m send(message)\n\u001b[0;32m    161\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[1;32m--> 162\u001b[0m     \u001b[39mawait\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mapp(scope, receive, _send)\n\u001b[0;32m    163\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mException\u001b[39;00m \u001b[39mas\u001b[39;00m exc:\n\u001b[0;32m    164\u001b[0m     request \u001b[39m=\u001b[39m Request(scope)\n",
      "File \u001b[1;32mc:\\Users\\vince\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\starlette\\middleware\\cors.py:83\u001b[0m, in \u001b[0;36mCORSMiddleware.__call__\u001b[1;34m(self, scope, receive, send)\u001b[0m\n\u001b[0;32m     80\u001b[0m origin \u001b[39m=\u001b[39m headers\u001b[39m.\u001b[39mget(\u001b[39m\"\u001b[39m\u001b[39morigin\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[0;32m     82\u001b[0m \u001b[39mif\u001b[39;00m origin \u001b[39mis\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n\u001b[1;32m---> 83\u001b[0m     \u001b[39mawait\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mapp(scope, receive, send)\n\u001b[0;32m     84\u001b[0m     \u001b[39mreturn\u001b[39;00m\n\u001b[0;32m     86\u001b[0m \u001b[39mif\u001b[39;00m method \u001b[39m==\u001b[39m \u001b[39m\"\u001b[39m\u001b[39mOPTIONS\u001b[39m\u001b[39m\"\u001b[39m \u001b[39mand\u001b[39;00m \u001b[39m\"\u001b[39m\u001b[39maccess-control-request-method\u001b[39m\u001b[39m\"\u001b[39m \u001b[39min\u001b[39;00m headers:\n",
      "File \u001b[1;32mc:\\Users\\vince\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\starlette\\middleware\\exceptions.py:79\u001b[0m, in \u001b[0;36mExceptionMiddleware.__call__\u001b[1;34m(self, scope, receive, send)\u001b[0m\n\u001b[0;32m     76\u001b[0m     handler \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_lookup_exception_handler(exc)\n\u001b[0;32m     78\u001b[0m \u001b[39mif\u001b[39;00m handler \u001b[39mis\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n\u001b[1;32m---> 79\u001b[0m     \u001b[39mraise\u001b[39;00m exc\n\u001b[0;32m     81\u001b[0m \u001b[39mif\u001b[39;00m response_started:\n\u001b[0;32m     82\u001b[0m     msg \u001b[39m=\u001b[39m \u001b[39m\"\u001b[39m\u001b[39mCaught handled exception, but response already started.\u001b[39m\u001b[39m\"\u001b[39m\n",
      "File \u001b[1;32mc:\\Users\\vince\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\starlette\\middleware\\exceptions.py:68\u001b[0m, in \u001b[0;36mExceptionMiddleware.__call__\u001b[1;34m(self, scope, receive, send)\u001b[0m\n\u001b[0;32m     65\u001b[0m     \u001b[39mawait\u001b[39;00m send(message)\n\u001b[0;32m     67\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[1;32m---> 68\u001b[0m     \u001b[39mawait\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mapp(scope, receive, sender)\n\u001b[0;32m     69\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mException\u001b[39;00m \u001b[39mas\u001b[39;00m exc:\n\u001b[0;32m     70\u001b[0m     handler \u001b[39m=\u001b[39m \u001b[39mNone\u001b[39;00m\n",
      "File \u001b[1;32mc:\\Users\\vince\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\fastapi\\middleware\\asyncexitstack.py:20\u001b[0m, in \u001b[0;36mAsyncExitStackMiddleware.__call__\u001b[1;34m(self, scope, receive, send)\u001b[0m\n\u001b[0;32m     18\u001b[0m     \u001b[39mexcept\u001b[39;00m \u001b[39mException\u001b[39;00m \u001b[39mas\u001b[39;00m e:\n\u001b[0;32m     19\u001b[0m         dependency_exception \u001b[39m=\u001b[39m e\n\u001b[1;32m---> 20\u001b[0m         \u001b[39mraise\u001b[39;00m e\n\u001b[0;32m     21\u001b[0m \u001b[39mif\u001b[39;00m dependency_exception:\n\u001b[0;32m     22\u001b[0m     \u001b[39m# This exception was possibly handled by the dependency but it should\u001b[39;00m\n\u001b[0;32m     23\u001b[0m     \u001b[39m# still bubble up so that the ServerErrorMiddleware can return a 500\u001b[39;00m\n\u001b[0;32m     24\u001b[0m     \u001b[39m# or the ExceptionMiddleware can catch and handle any other exceptions\u001b[39;00m\n\u001b[0;32m     25\u001b[0m     \u001b[39mraise\u001b[39;00m dependency_exception\n",
      "File \u001b[1;32mc:\\Users\\vince\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\fastapi\\middleware\\asyncexitstack.py:17\u001b[0m, in \u001b[0;36mAsyncExitStackMiddleware.__call__\u001b[1;34m(self, scope, receive, send)\u001b[0m\n\u001b[0;32m     15\u001b[0m scope[\u001b[39mself\u001b[39m\u001b[39m.\u001b[39mcontext_name] \u001b[39m=\u001b[39m stack\n\u001b[0;32m     16\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[1;32m---> 17\u001b[0m     \u001b[39mawait\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mapp(scope, receive, send)\n\u001b[0;32m     18\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mException\u001b[39;00m \u001b[39mas\u001b[39;00m e:\n\u001b[0;32m     19\u001b[0m     dependency_exception \u001b[39m=\u001b[39m e\n",
      "File \u001b[1;32mc:\\Users\\vince\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\starlette\\routing.py:718\u001b[0m, in \u001b[0;36mRouter.__call__\u001b[1;34m(self, scope, receive, send)\u001b[0m\n\u001b[0;32m    716\u001b[0m \u001b[39mif\u001b[39;00m match \u001b[39m==\u001b[39m Match\u001b[39m.\u001b[39mFULL:\n\u001b[0;32m    717\u001b[0m     scope\u001b[39m.\u001b[39mupdate(child_scope)\n\u001b[1;32m--> 718\u001b[0m     \u001b[39mawait\u001b[39;00m route\u001b[39m.\u001b[39mhandle(scope, receive, send)\n\u001b[0;32m    719\u001b[0m     \u001b[39mreturn\u001b[39;00m\n\u001b[0;32m    720\u001b[0m \u001b[39melif\u001b[39;00m match \u001b[39m==\u001b[39m Match\u001b[39m.\u001b[39mPARTIAL \u001b[39mand\u001b[39;00m partial \u001b[39mis\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n",
      "File \u001b[1;32mc:\\Users\\vince\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\starlette\\routing.py:276\u001b[0m, in \u001b[0;36mRoute.handle\u001b[1;34m(self, scope, receive, send)\u001b[0m\n\u001b[0;32m    274\u001b[0m     \u001b[39mawait\u001b[39;00m response(scope, receive, send)\n\u001b[0;32m    275\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[1;32m--> 276\u001b[0m     \u001b[39mawait\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mapp(scope, receive, send)\n",
      "File \u001b[1;32mc:\\Users\\vince\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\starlette\\routing.py:66\u001b[0m, in \u001b[0;36mrequest_response.<locals>.app\u001b[1;34m(scope, receive, send)\u001b[0m\n\u001b[0;32m     64\u001b[0m request \u001b[39m=\u001b[39m Request(scope, receive\u001b[39m=\u001b[39mreceive, send\u001b[39m=\u001b[39msend)\n\u001b[0;32m     65\u001b[0m \u001b[39mif\u001b[39;00m is_coroutine:\n\u001b[1;32m---> 66\u001b[0m     response \u001b[39m=\u001b[39m \u001b[39mawait\u001b[39;00m func(request)\n\u001b[0;32m     67\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[0;32m     68\u001b[0m     response \u001b[39m=\u001b[39m \u001b[39mawait\u001b[39;00m run_in_threadpool(func, request)\n",
      "File \u001b[1;32mc:\\Users\\vince\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\fastapi\\routing.py:274\u001b[0m, in \u001b[0;36mget_request_handler.<locals>.app\u001b[1;34m(request)\u001b[0m\n\u001b[0;32m    272\u001b[0m     \u001b[39mraise\u001b[39;00m RequestValidationError(_normalize_errors(errors), body\u001b[39m=\u001b[39mbody)\n\u001b[0;32m    273\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[1;32m--> 274\u001b[0m     raw_response \u001b[39m=\u001b[39m \u001b[39mawait\u001b[39;00m run_endpoint_function(\n\u001b[0;32m    275\u001b[0m         dependant\u001b[39m=\u001b[39mdependant, values\u001b[39m=\u001b[39mvalues, is_coroutine\u001b[39m=\u001b[39mis_coroutine\n\u001b[0;32m    276\u001b[0m     )\n\u001b[0;32m    278\u001b[0m     \u001b[39mif\u001b[39;00m \u001b[39misinstance\u001b[39m(raw_response, Response):\n\u001b[0;32m    279\u001b[0m         \u001b[39mif\u001b[39;00m raw_response\u001b[39m.\u001b[39mbackground \u001b[39mis\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n",
      "File \u001b[1;32mc:\\Users\\vince\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\fastapi\\routing.py:191\u001b[0m, in \u001b[0;36mrun_endpoint_function\u001b[1;34m(dependant, values, is_coroutine)\u001b[0m\n\u001b[0;32m    188\u001b[0m \u001b[39massert\u001b[39;00m dependant\u001b[39m.\u001b[39mcall \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m, \u001b[39m\"\u001b[39m\u001b[39mdependant.call must be a function\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[0;32m    190\u001b[0m \u001b[39mif\u001b[39;00m is_coroutine:\n\u001b[1;32m--> 191\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mawait\u001b[39;00m dependant\u001b[39m.\u001b[39mcall(\u001b[39m*\u001b[39m\u001b[39m*\u001b[39mvalues)\n\u001b[0;32m    192\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[0;32m    193\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mawait\u001b[39;00m run_in_threadpool(dependant\u001b[39m.\u001b[39mcall, \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mvalues)\n",
      "File \u001b[1;32mc:\\SMU\\Y3T1\\IS212 Software Project Management\\spm-g5t3\\pytest\\..\\api\\application.py:148\u001b[0m, in \u001b[0;36mapplication\u001b[1;34m(application)\u001b[0m\n\u001b[0;32m    144\u001b[0m \u001b[39m@app\u001b[39m\u001b[39m.\u001b[39mpost(\u001b[39m\"\u001b[39m\u001b[39m/api/application\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[0;32m    145\u001b[0m \u001b[39m@router\u001b[39m\u001b[39m.\u001b[39mpost(\u001b[39m\"\u001b[39m\u001b[39m/api/application\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[0;32m    146\u001b[0m \u001b[39masync\u001b[39;00m \u001b[39mdef\u001b[39;00m \u001b[39mapplication\u001b[39m(application: PostApplication):\n\u001b[0;32m    147\u001b[0m     response \u001b[39m=\u001b[39m (\n\u001b[1;32m--> 148\u001b[0m         supabase\u001b[39m.\u001b[39;49mtable(\u001b[39m\"\u001b[39;49m\u001b[39mapplication\u001b[39;49m\u001b[39m\"\u001b[39;49m)\u001b[39m.\u001b[39;49minsert(application\u001b[39m.\u001b[39;49mdict())\u001b[39m.\u001b[39;49mexecute()\n\u001b[0;32m    149\u001b[0m     )\n\u001b[0;32m    150\u001b[0m     \u001b[39mif\u001b[39;00m response[\u001b[39m\"\u001b[39m\u001b[39mstatus_code\u001b[39m\u001b[39m\"\u001b[39m] \u001b[39m==\u001b[39m \u001b[39m200\u001b[39m:\n\u001b[0;32m    151\u001b[0m         \u001b[39mreturn\u001b[39;00m {\u001b[39m\"\u001b[39m\u001b[39msuccess\u001b[39m\u001b[39m\"\u001b[39m: \u001b[39mTrue\u001b[39;00m, \u001b[39m\"\u001b[39m\u001b[39mdata\u001b[39m\u001b[39m\"\u001b[39m: response[\u001b[39m'\u001b[39m\u001b[39mdata\u001b[39m\u001b[39m'\u001b[39m]}\n",
      "File \u001b[1;32mc:\\Users\\vince\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\postgrest\\_sync\\request_builder.py:70\u001b[0m, in \u001b[0;36mSyncQueryRequestBuilder.execute\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m     68\u001b[0m         \u001b[39mreturn\u001b[39;00m APIResponse[_ReturnT]\u001b[39m.\u001b[39mfrom_http_request_response(r)\n\u001b[0;32m     69\u001b[0m     \u001b[39melse\u001b[39;00m:\n\u001b[1;32m---> 70\u001b[0m         \u001b[39mraise\u001b[39;00m APIError(r\u001b[39m.\u001b[39mjson())\n\u001b[0;32m     71\u001b[0m \u001b[39mexcept\u001b[39;00m ValidationError \u001b[39mas\u001b[39;00m e:\n\u001b[0;32m     72\u001b[0m     \u001b[39mraise\u001b[39;00m APIError(r\u001b[39m.\u001b[39mjson()) \u001b[39mfrom\u001b[39;00m \u001b[39me\u001b[39;00m\n",
      "\u001b[1;31mAPIError\u001b[0m: {'code': '23505', 'details': 'Key (application_id)=(69) already exists.', 'hint': None, 'message': 'duplicate key value violates unique constraint \"application_pkey\"'}"
     ]
    }
   ],
   "source": [
    "from fastapi.testclient import TestClient\n",
    "import sys\n",
    "\n",
    "sys.path.append(\"..\")\n",
    "from main import app\n",
    "client = TestClient(app)\n",
    "\n",
    "# response = client.get(\"/api/application?application_id=1\")\n",
    "response = client.post(\n",
    "        \"/api/application\",\n",
    "        json={\n",
    "            \"staff_id\": 140944,\n",
    "            \"listing_id\": 1,\n",
    "            \"application_reason\": \"test_application.py\"\n",
    "        }\n",
    "    )\n",
    "print(response.json())\n",
    "print(response.status_code)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Empty dict is false\n"
     ]
    }
   ],
   "source": [
    "if {}:\n",
    "    print(\"Empty dict is true\")\n",
    "else:\n",
    "    print(\"Empty dict is false\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
